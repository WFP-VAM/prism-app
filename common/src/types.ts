export type BBOX = Readonly<[number, number, number, number]>;

export type LAYER = {
  id: Readonly<string>;
  name: Readonly<string>;
  description: Readonly<string>;
  tags: Readonly<string[]>;
  bbox: Readonly<[number, number, number, number]>;
};

export type MakeOptional<Type, Key extends keyof Type> = Omit<Type, Key> &
  Partial<Pick<Type, Key>>;
