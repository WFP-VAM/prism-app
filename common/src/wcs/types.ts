export type CoverageFormat =
  | "ArcGrid"
  | "ArcGrid-GZIP"
  | "bmp"
  | "jpeg"
  | "geotiff"
  | "gif"
  | "png"
  | "tif";
