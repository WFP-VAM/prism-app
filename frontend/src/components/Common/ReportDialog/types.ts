import { Style } from '@react-pdf/types';
import { TFunctionResult } from 'i18next';

export interface PDFLegendDefinition {
  value: string | number | TFunctionResult;
  style: Style | Style[];
}

// This numbers depends on table header and row height
export const MAX_TABLE_ROWS_PER_PAGE = 29;
export const FIRST_PAGE_TABLE_ROWS = 9;
