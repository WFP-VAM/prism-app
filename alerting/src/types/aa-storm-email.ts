import { WindState } from 'prism-common';

export type LastStates = {
  [stormName: string]: {
    status: WindState;
    refTime: string;
  };
};

export interface ShortReport {
  ref_time: string;
  state: WindState;
  path: string;
}

export interface ShortReportsResponseBody {
  [date: string]: {
    [stormName: string]: ShortReport[];
  };
}
