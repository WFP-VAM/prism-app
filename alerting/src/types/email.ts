export interface StormAlertData {
    email: string,
    cycloneName: string,
    cycloneTime: string,
    activatedTriggers?: {
        districts48kt: string[],
        districts64kt: string[],
        windspeed: string,
    },
    redirectUrl: string,
    base64Image: string,
    readiness: boolean,
}

export interface StormAlertEmail extends Omit<StormAlertData, 'email' | 'activatedTriggers' | 'base64Image'> {
    alertTitle: string;
    unsubscribeUrl: string;
    activatedTriggers?: {
        districts48kt: string,
        districts64kt: string,
        windspeed: string,
    },
}