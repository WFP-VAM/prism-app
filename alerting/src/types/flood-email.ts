export const TRIGGER_STATUSES = [
  'not exceeded',
  'bankfull',
  'moderate',
  'severe',
] as const;

export type TriggerStatus = (typeof TRIGGER_STATUSES)[number];

export interface FloodAlertEmailData {
  email: string | string[];
  title: string; // e.g., "Flood trigger activated: Moderate"
  triggerStatus: TriggerStatus;
  date: string; // ISO date
  stations?: {
    name: string;
    id?: number | string;
    river?: string;
  }[];
  stationsByStatus?: Record<string, string[]>; // Groups stations by their trigger status
  redirectUrl: string;
  base64Image: string;
}
