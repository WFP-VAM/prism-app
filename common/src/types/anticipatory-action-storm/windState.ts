export enum WindState {
  monitoring = "monitoring",
  ready = "ready",
  activated_48kt = "activated_48kt",
  activated_64kt = "activated_64kt",
}

export const displayWindState: Partial<Record<WindState, string>> = {
  [WindState.activated_64kt]: "> 118 km/h",
  [WindState.activated_48kt]: "> 89 km/h",
};
