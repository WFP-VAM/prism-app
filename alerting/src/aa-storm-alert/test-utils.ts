import { LandfallInfo } from 'prism-common/dist/types/anticipatory-action-storm/reportResponse';
import { WindState } from 'prism-common/dist/types/anticipatory-action-storm/windState';

export function buildLandfallInfo({
  landfall_time = ['2025-01-13 06:00:00', '2025-01-13 18:00:00'],
}: Partial<LandfallInfo>): LandfallInfo {
  return {
    landfall_time,
    landfall_impact_district: 'Mogincual',
    landfall_impact_intensity: [],
    landfall_leadtime_hours: [0.0, 12.0],
    is_coastal: true,
  };
}

export function buildDetailedReport({
  landfall_detected = false,
  status = WindState.monitoring,
  affected48ktDistrict = [],
  affected64ktDistrict = [],
  landfallInfo = {} as LandfallInfo | Record<string, never>,
}: {
  landfall_detected?: boolean;
  status?: WindState;
  affected48ktDistrict?: string[];
  affected64ktDistrict?: string[];
  landfallInfo?: LandfallInfo | Record<string, never>;
}) {
  return {
    forecast_details: {
      cyclone_name: 'ELVIS',
      season: 20242025,
      reference_time: '2025-01-31T06:00:00Z',
      basin: 'SWI',
    },
    landfall_detected,
    landfall_info: landfallInfo,
    ready_set_results: {
      status,
      exposed_area_48kt: {
        affected_districts: affected48ktDistrict,
        polygon: {
          type: 'Polygon',
          coordinates: [],
        },
      },
      exposed_area_64kt: {
        affected_districts: affected64ktDistrict,
        polygon: {
          type: 'Polygon',
          coordinates: [],
        },
      },
    },
    uncertainty_cone: {
      type: 'Polygon',
      /* coordinates are not filled in this builder because not usefull */
      coordinates: [],
    },
    /* timeSeries are not filled in this builder because not usefull */
    time_series: {},
  };
}

export function buildAnticipatoryActionAlerts({
  id = 1,
  country = 'mozambique',
  emails = [],
  prismUrl = 'http://example.com',
  lastRanAt = new Date(),
  lastStates = {},
}) {
  return {
    id,
    country,
    emails,
    prismUrl,
    lastRanAt,
    lastStates,
  };
}
