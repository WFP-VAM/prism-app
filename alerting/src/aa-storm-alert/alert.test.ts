jest.mock('node-fetch');
import nodeFetch from 'node-fetch';
import {
  buildEmailPayloads,
  filterOutAlreadyProcessedReports,
  getLatestAvailableReports,
} from './alert';
import { buildDetailedReport, buildLandfallInfo } from './test-utils';
import { WindState } from 'prism-common/dist/types/anticipatory-action-storm/windState';
import moment from 'moment';
import { LastStates } from '../types/aa-storm-email';

describe('alert mechanism', () => {
  describe('getLatestAvailableReports()', () => {
    const mockedFetch = nodeFetch as unknown as jest.Mock;
    afterEach(() => {
      jest.resetAllMocks();
    });

    const tests = [
      {
        description: 'when short reports have one storm',
        data: {
          '2025-01-31': {
            elvis: [
              {
                ref_time: '2025-01-31T06:00:00Z',
                state: 'monitoring',
                path: 'elvis/2025-01-31T06:00:00Z.json',
              },
            ],
          },
        },
        expected: [
          {
            ref_time: '2025-01-31T06:00:00Z',
            state: 'monitoring',
            path: 'elvis/2025-01-31T06:00:00Z.json',
          },
        ],
      },
      {
        description: 'when short reports have 2 storms',
        data: {
          '2025-01-30': {
            '07-20242025': [
              {
                ref_time: '2025-01-30T06:00:00Z',
                state: 'monitoring',
                path: '07-20242025/2025-01-30T06:00:00Z.json',
              },
              {
                ref_time: '2025-01-30T12:00:00Z',
                state: 'monitoring',
                path: '07-20242025/2025-01-30T12:00:00Z.json',
              },
            ],
            elvis: [
              {
                ref_time: '2025-01-30T06:00:00Z',
                state: 'monitoring',
                path: 'elvis/2025-01-30T06:00:00Z.json',
              },
              {
                ref_time: '2025-01-30T18:00:00Z',
                state: 'monitoring',
                path: 'elvis/2025-01-30T18:00:00Z.json',
              },
            ],
          },
        },
        expected: [
          {
            ref_time: '2025-01-30T12:00:00Z',
            state: 'monitoring',
            path: '07-20242025/2025-01-30T12:00:00Z.json',
          },
          {
            ref_time: '2025-01-30T18:00:00Z',
            state: 'monitoring',
            path: 'elvis/2025-01-30T18:00:00Z.json',
          },
        ],
      },
    ];
    it.each(tests)(
      'get latest reports $description',
      async ({ data, expected }) => {
        mockedFetch.mockResolvedValue({ json: () => data });

        const result = await getLatestAvailableReports();
        expect(result).toEqual(expected);
      },
    );

    it('it returns an empty array when request fails', async () => {
      mockedFetch.mockRejectedValue(null);
      const result = await getLatestAvailableReports();
      expect(result).toEqual([]);
    });
  });

  describe('buildEmailPayloads()', () => {
    const mockedFetch = nodeFetch as unknown as jest.Mock;
    afterEach(() => {
      jest.resetAllMocks();
    });

    const tests = [
      {
        description:
          'returns a payload when detailed report indicates that readiness is triggered',
        shortReports: [
          {
            ref_time: '2025-01-31T06:00:00Z',
            state: WindState.ready,
            path: 'elvis/2025-01-31T06:00:00Z.json',
          },
        ],
        data: buildDetailedReport({
          status: WindState.ready,
        }),
      },
      {
        description:
          'returns a payload when detailed report indicates that activation48 is triggered and pilot activated districts for 48kt winds are parts of the exposed districts',
        shortReports: [
          {
            ref_time: '2025-01-31T06:00:00Z',
            state: WindState.activated_48kt,
            path: 'elvis/2025-01-31T06:00:00Z.json',
          },
        ],
        data: buildDetailedReport({
          status: WindState.activated_48kt,
          affected48ktDistrict: ['Angoche'],
        }),
      },
      {
        description:
          'does not return a payload when detailed report indicates that activation48 is triggered but there is no pilot activated districts for 48kt exposed',
        shortReports: [
          {
            ref_time: '2025-01-31T06:00:00Z',
            state: WindState.activated_48kt,
            path: 'elvis/2025-01-31T06:00:00Z.json',
          },
        ],
        data: buildDetailedReport({
          status: WindState.activated_48kt,
          affected48ktDistrict: [],
        }),
      },
      {
        description:
          'returns a payload when detailed report indicates that activation64 is triggered and pilot activated districts for 64kt winds are parts of the exposed districts',
        shortReports: [
          {
            ref_time: '2025-01-31T06:00:00Z',
            state: WindState.activated_64kt,
            path: 'elvis/2025-01-31T06:00:00Z.json',
          },
        ],
        data: buildDetailedReport({
          status: WindState.activated_64kt,
          affected64ktDistrict: ['Namacurra'],
        }),
      },
      {
        description:
          'does not return a payload when detailed report indicates that activation64 is triggered but there is no pilot activated districts for 64kt exposed',
        shortReports: [
          {
            ref_time: '2025-01-31T06:00:00Z',
            state: WindState.activated_64kt,
            path: 'elvis/2025-01-31T06:00:00Z.json',
          },
        ],
        data: buildDetailedReport({
          status: WindState.activated_64kt,
          affected64ktDistrict: [],
        }),
      },
      {
        description:
          'returns no payload when detailed report indicates that activation is triggered but landfall already occured',
        shortReports: [
          {
            ref_time: '2025-01-31T06:00:00Z',
            state: WindState.activated_64kt,
            path: 'elvis/2025-01-31T06:00:00Z.json',
          },
        ],
        data: buildDetailedReport({
          landfall_detected: false,
          status: WindState.activated_64kt,
          affected64ktDistrict: ['Namacurra'],
          landfallInfo: buildLandfallInfo({
            landfall_time: [
              '2025-01-13 06:00:00',
              moment().subtract(1, 'hour').format('YYYY-MM-DD HH:mm:ss'), // now - 1 hour; ie landfall occured already
            ],
          }),
        }),
      },
    ];
    it.each(tests)('$description', async ({ data, shortReports }) => {
      mockedFetch.mockResolvedValue({ json: () => data });

      const emailPayloads = await buildEmailPayloads(
        shortReports,
        'https://prism.wfp.org',
        ['test@test.com'],
      );
      expect(emailPayloads).toMatchSnapshot();
    });
  });

  describe('filterOutAlreadyProcessedReports()', () => {
    const tests = [
      {
        description:
          'when there is already processed report stored in the db which is older than available report',
        availableReports: [
          {
            ref_time: '2025-02-09T18:00:00Z',
            state: WindState.monitoring,
            path: 'vince/2025-02-09T18:00:00Z.json',
          },
        ],
        lastStates: {
          vince: {
            status: WindState.monitoring,
            refTime: '2025-02-09T12:00:00Z',
          },
        },
      },
      {
        description:
          'when there is already processed report stored in the db which is as old as available report',
        availableReports: [
          {
            ref_time: '2025-02-09T18:00:00Z',
            state: WindState.monitoring,
            path: 'vince/2025-02-09T18:00:00Z.json',
          },
        ],
        lastStates: {
          vince: {
            status: WindState.monitoring,
            refTime: '2025-02-09T18:00:00Z',
          },
        },
      },
      {
        description:
          'when there is no already processed report stored in the db',
        availableReports: [
          {
            ref_time: '2025-02-09T18:00:00Z',
            state: WindState.monitoring,
            path: 'vince/2025-02-09T18:00:00Z.json',
          },
        ],
        lastStates: undefined,
      },
      {
        description: 'when there are several available reports',
        availableReports: [
          {
            ref_time: '2025-02-09T18:00:00Z',
            state: WindState.monitoring,
            path: 'vince/2025-02-09T18:00:00Z.json',
          },
          {
            ref_time: '2025-02-09T18:00:00Z',
            state: WindState.monitoring,
            path: 'chido/2025-02-09T18:00:00Z.json',
          },
        ],
        lastStates: {
          vince: {
            status: WindState.monitoring,
            refTime: '2025-02-09T12:00:00Z',
          },
          chido: {
            status: WindState.monitoring,
            refTime: '2025-02-09T12:00:00Z',
          },
        },
      },
    ];
    it.each(tests)(
      'returns filtered available reports $description',
      ({ availableReports, lastStates }) => {
        const filteredReports = filterOutAlreadyProcessedReports(
          availableReports,
          lastStates as LastStates,
        );
        expect(filteredReports).toMatchSnapshot();
      },
    );
  });
});
