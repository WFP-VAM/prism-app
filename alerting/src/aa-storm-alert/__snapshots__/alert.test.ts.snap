// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alert mechanism buildEmailPayloads() does not return a payload when detailed report indicates that activation48 is triggered but there is no pilot activated districts for 48kt exposed 1`] = `[]`;

exports[`alert mechanism buildEmailPayloads() does not return a payload when detailed report indicates that activation64 is triggered but there is no pilot activated districts for 64kt exposed 1`] = `[]`;

exports[`alert mechanism buildEmailPayloads() returns a payload when detailed report indicates that activation48 is triggered and pilot activated districts for 48kt winds are parts of the exposed districts 1`] = `
[
  {
    "activatedTriggers": {
      "districts48kt": [
        "Angoche",
      ],
      "districts64kt": [],
    },
    "base64Image": "",
    "cycloneName": "ELVIS",
    "cycloneTime": "2025-01-31T06:00:00Z",
    "email": [
      "test@test.com",
    ],
    "redirectUrl": "https://prism.wfp.org/?hazardLayerIds=anticipatory_action_storm&date=2025-01-31",
    "status": "activated_48kt",
  },
]
`;

exports[`alert mechanism buildEmailPayloads() returns a payload when detailed report indicates that activation64 is triggered and pilot activated districts for 64kt winds are parts of the exposed districts 1`] = `
[
  {
    "activatedTriggers": {
      "districts48kt": [],
      "districts64kt": [
        "Namacurra",
      ],
    },
    "base64Image": "",
    "cycloneName": "ELVIS",
    "cycloneTime": "2025-01-31T06:00:00Z",
    "email": [
      "test@test.com",
    ],
    "redirectUrl": "https://prism.wfp.org/?hazardLayerIds=anticipatory_action_storm&date=2025-01-31",
    "status": "activated_64kt",
  },
]
`;

exports[`alert mechanism buildEmailPayloads() returns a payload when detailed report indicates that readiness is triggered 1`] = `
[
  {
    "activatedTriggers": {
      "districts48kt": [],
      "districts64kt": [],
    },
    "base64Image": "",
    "cycloneName": "ELVIS",
    "cycloneTime": "2025-01-31T06:00:00Z",
    "email": [
      "test@test.com",
    ],
    "redirectUrl": "https://prism.wfp.org/?hazardLayerIds=anticipatory_action_storm&date=2025-01-31",
    "status": "ready",
  },
]
`;

exports[`alert mechanism buildEmailPayloads() returns no payload when detailed report indicates that activation is triggered but landfall already occured 1`] = `[]`;

exports[`alert mechanism filterOutAlreadyProcessedReports() returns filtered available reports when there are several available reports 1`] = `
[
  {
    "path": "vince/2025-02-09T18:00:00Z.json",
    "ref_time": "2025-02-09T18:00:00Z",
    "state": "monitoring",
  },
  {
    "path": "chido/2025-02-09T18:00:00Z.json",
    "ref_time": "2025-02-09T18:00:00Z",
    "state": "monitoring",
  },
]
`;

exports[`alert mechanism filterOutAlreadyProcessedReports() returns filtered available reports when there is already processed report stored in the db which is as old as available report 1`] = `[]`;

exports[`alert mechanism filterOutAlreadyProcessedReports() returns filtered available reports when there is already processed report stored in the db which is older than available report 1`] = `
[
  {
    "path": "vince/2025-02-09T18:00:00Z.json",
    "ref_time": "2025-02-09T18:00:00Z",
    "state": "monitoring",
  },
]
`;

exports[`alert mechanism filterOutAlreadyProcessedReports() returns filtered available reports when there is no already processed report stored in the db 1`] = `
[
  {
    "path": "vince/2025-02-09T18:00:00Z.json",
    "ref_time": "2025-02-09T18:00:00Z",
    "state": "monitoring",
  },
]
`;
